### PR TITLE
Log file metrics

### PR DESCRIPTION
**Pull request by AndrewClugston**
Gathers log file metrics including:
-Log file read time
-Log file replay time
-Distinct sym count
-Checks records are successfully appended
-Outputs number of appended records

### CODE REVIEW

File: code/common/subscriptions.q [hunk 1]
```diff
@@ -10,22 +9,23 @@ SUBSCRIPTIONS:([]procname:`symbol$();proctype:`symbol$();w:`int$();table:();inst
...
 updatesubscriptions:{[proc;tab;instrs]
 	/-delete any inactive subscriptions
 	delete from `.sub.SUBSCRIPTIONS where not active;
 	if[instrs~`;instrs,:()];
 	.sub.SUBSCRIPTIONS::0!(4!SUBSCRIPTIONS)upsert enlist proc,`table`instruments`createdtime`active!(tab;instrs;.proc.cp[];1b);
-	}	
+	}
 
 reconnectinit:0b;		//has the reconnect custom function been initialised
+chksym:{[x]?[x;();1b;enlist[`sym]!enlist[`sym]]}
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: ``{x!x}(),`sym``

File: code/common/subscriptions.q [hunk 2]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
+                        /-the second element of details is a pair (log count;logfile)
+                        recordsappended:@[{-11!x};details 1;{.lg.e[`subscribe;"could not replay the log file: ", x]}];
+                        /-calculate difference between start and end of replay
+                        .lg.o[`ReplayTimer;"Replay Timer end: ", string[replayEnd:.z.n]];
+                        replayTimer-:@[{x};replayEnd;{.lg.e[`ReplayTimer;"Error calculating timer: ", replayEnd]}];
+                        /-display result
+                        .lg.o[`ReplayTimer;"Replay Timer total time taken: ", string[neg replayTimer]];
+                        $[logfilecount~recordsappended;.lg.o[`subscribe;"all ",string[recordsappended]," records in the logfile are succesfully appended"];.lg.o[`subscribe;"all the records in the logfile are not succesfully appended, only ",string[recordsappended]," records have been appended"]];
+                        /-reset the upd function back to original upd
+                        @[`.;`upd;:;origupd]];
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: 82-92 is indented by an extra space relative to 81.

Closing `]` should be on newline.

File: code/common/subscriptions.q [hunk 3]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
 				.lg.o[`subscribe;"using the .sub.replayupd function as not replaying all tables or instruments"];
 				@[`.;`upd;:;.sub.replayupd[origupd;subtabs;instrs]]];
-			/-the second element of details is a pair (log count;logfile)
-			@[{-11!x;};details 1;{.lg.e[`subscribe;"could not replay the log file: ", x]}];
-			/-reset the upd function back to original upd
-			@[`.;`upd;:;origupd]];
+			logfilecount:(details[1])[0];
+			/-check log replay time
+			/-display replaying log file start time
+                        .lg.o[`ReplayTimer;"Replay Timer start: ",string[replayTimer:.z.n]];
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: No need for square brackets with `}`

File: code/common/subscriptions.q [hunk 4]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
+                        /-the second element of details is a pair (log count;logfile)
+                        recordsappended:@[{-11!x};details 1;{.lg.e[`subscribe;"could not replay the log file: ", x]}];
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: Whitespace.

File: code/common/subscriptions.q [hunk 5]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
+			/-check log read time
+                        /-display read log file start time
+                        .lg.o[`ReadTimer;"Read Timer start: ",string[readTimer:.z.n]];
+                        /-read log file
+                        @[{-11!(-2;x);};first 1_details 1;{.lg.e[`ReadTimer;"Log read error with output: ", x]}];
+                        /-calculate difference between start and end of read
+                        .lg.o[`ReadTimer;"Read Timer end: ", string[readEnd:.z.n]];
+                        readTimer-:@[{x};.z.n;{.lg.e[`ReadTimer;"Error calculating timer: ", readEnd]}];
+                        /-display result
+                        .lg.o[`ReadTimer;"Read Timer total time taken: ", string[neg readTimer]];
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: More weird indenting in this section + it seems there are a lot of similarities between read/replay, look for ways to leverage the same code for both.

File: code/common/subscriptions.q [hunk 6]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
 		updatesubscriptions[proc;;instrs]each subtabs];
+			
+		 	.lg.o[`DistinctSym;"Total distinct syms: ",string count distinct raze chksym each tables[]];	
+			/-check log read time
+                        /-display read log file start time
+                        .lg.o[`ReadTimer;"Read Timer start: ",string[readTimer:.z.n]];
+                        /-read log file
+                        @[{-11!(-2;x);};first 1_details 1;{.lg.e[`ReadTimer;"Log read error with output: ", x]}];
+                        /-calculate difference between start and end of read
+                        .lg.o[`ReadTimer;"Read Timer end: ", string[readEnd:.z.n]];
```
Comment (ThomasSmyth) [inline code/common/subscriptions.q]: Remove excess squares.

File: code/common/subscriptions.q [hunk 7]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
-			/-set the replayupd function to be upd globally 
+			/-set the replayupd function to be upd globally
 			if[not (tabs;instrs)~(`;`);
 				.lg.o[`subscribe;"using the .sub.replayupd function as not replaying all tables or instruments"];
 				@[`.;`upd;:;.sub.replayupd[origupd;subtabs;instrs]]];
-			/-the second element of details is a pair (log count;logfile)
-			@[{-11!x;};details 1;{.lg.e[`subscribe;"could not replay the log file: ", x]}];
-			/-reset the upd function back to original upd
-			@[`.;`upd;:;origupd]];
+			logfilecount:(details[1])[0];
```
Comment (claire-sloan) [inline code/common/subscriptions.q]: can probably remove some brackets here

File: code/common/subscriptions.q [hunk 8]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
+			/-check log replay time
+			/-display replaying log file start time
+                        .lg.o[`ReplayTimer;"Replay Timer start: ",string[replayTimer:.z.n]];
+                        /-the second element of details is a pair (log count;logfile)
+                        recordsappended:@[{-11!x};details 1;{.lg.e[`subscribe;"could not replay the log file: ", x]}];
+                        /-calculate difference between start and end of replay
+                        .lg.o[`ReplayTimer;"Replay Timer end: ", string[replayEnd:.z.n]];
```
Comment (claire-sloan) [inline code/common/subscriptions.q]: whitespace here and next few lines 
can remove []

File: code/common/subscriptions.q [hunk 9]
```diff
@@ -72,17 +72,41 @@ subscribe:{[tabs;instrs;setschema;replaylog;proc]
...
+                        replayTimer-:@[{x};replayEnd;{.lg.e[`ReplayTimer;"Error calculating timer: ", replayEnd]}];
+                        /-display result
+                        .lg.o[`ReplayTimer;"Replay Timer total time taken: ", string[neg replayTimer]];
+                        $[logfilecount~recordsappended;.lg.o[`subscribe;"all ",string[recordsappended]," records in the logfile are succesfully appended"];.lg.o[`subscribe;"all the records in the logfile are not succesfully appended, only ",string[recordsappended]," records have been appended"]];
```
Comment (claire-sloan) [inline code/common/subscriptions.q]: too long split over a few lines

**Comment by ThomasSmyth** [comment]
You never added this "closes #12" to the description.

Why is that not part of the release?

**Pull request merged**